File: tests/call_native.c
// Test the CALL_NATIVE instruction.
//
// (c) Reuben Thomas 1995-2018
//
// The package is distributed under the GNU Public License version 3, or,
// at your option, any later version.
//
// THIS PROGRAM IS PROVIDED AS IS, WITH NO WARRANTY. USE IS AT THE USER‘S
// RISK.

#include "tests.h"


int exception = 0;
WORD temp;


#define CORRECT 37

static void test(state *S)
{
    PUSH(CORRECT);
}

int main(void)
{
    state *S = init_default_stacks(4096);

    ass_native_pointer(S, test); ass_action(S, O_CALL_NATIVE); ass_number(S, 0);
    ass_action(S, O_HALT);

    WORD res = run(S);
    if (res != 0) {
        printf("Error in CALL_NATIVE tests: test aborted with return code %"PRI_WORD"\n", res);
        exit(1);
    }

    WORD top_word;
    load_stack(S->S0, S->SDEPTH, 0, &top_word);
    printf("Top of stack is %"PRI_WORD"; should be %d\n", top_word, CORRECT);
    show_data_stack(S);
    if (top_word != CORRECT) {
        printf("Error in CALL_NATIVE tests: incorrect value on top of stack\n");
        exit(1);
    }

    destroy(S);

    assert(exception == 0);
    printf("CALL_NATIVE tests ran OK\n");
    return 0;
}

File: tests/call_native.py
# Test the CALL_NATIVE instruction.
#
# (c) Reuben Thomas 1995-2018
#
# The package is distributed under the GNU Public License version 3, or,
# at your option, any later version.
#
# THIS PROGRAM IS PROVIDED AS IS, WITH NO WARRANTY. USE IS AT THE USER‘S
# RISK.

from smite import *
VM = State()
VM.globalize(globals())


correct = 37

# Test callback
def test(state):
    S.push(correct)
test_func = SMITE_CALLBACK_FUNCTYPE(test)

# Test code
pointer(cast(test_func, c_void_p).value)
action(CALL_NATIVE)
number(0)
action(HALT)

# Test
res = run()
if res != 0:
    print("Error in CALL_NATIVE tests: test aborted with return code {}".format(res))
    sys.exit(1)

top_word = S.pop()
print("Top of stack is {}; should be {}".format(top_word, correct))
print("Data stack: {}".format(S))
if top_word != correct:
    print("Error in CALL_NATIVE tests: incorrect value on top of stack\n")
    sys.exit(1)

print("CALL_NATIVE tests ran OK")
